### PR TITLE
Use upstream `DataType::from_str` in arrow-cast

### PR DESCRIPTION
**Pull request by alamb**
~Draft until https://github.com/apache/arrow-rs/pull/5994 is released (likely `52.2.0`)~

## Which issue does this PR close?

N/A

## Rationale for this change

@samuelcolvin  ported this code upstream to arrow-rs in https://github.com/apache/arrow-rs/pull/5994 so let's use that code

## What changes are included in this PR?

Remove DataType parsing code in DataFusion and use upstreamed copy

### CODE REVIEW

File: datafusion/functions/src/core/arrow_cast.rs [hunk 1]
```diff
@@ -139,767 +137,9 @@ fn data_type_from_args(args: &[Expr]) -> Result<DataType> {
...
-                        message.contains(expected_message),
-                        "\n\ndid not find expected in actual.\n\nexpected: {expected_message}\nactual:{message}\n"
-                    );
-                    // errors should also contain  a help message
-                    assert!(message.contains("Must be a supported arrow type name such as 'Int32' or 'Timestamp(Nanosecond, None)'"));
-                }
-            }
-        }
-    }
+    val.parse().map_err(|e| match e {
```
Comment (alamb) [inline datafusion/functions/src/core/arrow_cast.rs:141]: Simply use the upstream impl now

**Comment by alamb** [comment]
I also double checked that it has support for  `BinaryView` and `Utf8View`: https://github.com/apache/arrow-rs/blob/80ed7128510bac114c6feec08c34ef3beed3a44a/arrow-schema/src/datatype_parse.rs#L443-L445

File: datafusion/functions/src/core/arrow_cast.rs [hunk 2]
```diff
@@ -44,7 +42,7 @@ use datafusion_expr::{ColumnarValue, Expr, ScalarUDFImpl, Signature, Volatility}
 /// select cast(column_x as int) ...
 /// ```
 ///
-/// You can use the `arrow_cast` functiont to cast to a specific arrow type
+/// You can use the `arrow_cast` function to cast to a specific arrow type
```
Comment (comphead) [inline datafusion/functions/src/core/arrow_cast.rs:45]:
```suggestion
/// Use the `arrow_cast` function to cast to a specific arrow type
```

File: datafusion/functions/src/core/arrow_cast.rs [hunk 3]
```diff
@@ -139,767 +137,9 @@ fn data_type_from_args(args: &[Expr]) -> Result<DataType> {
...
-                        "\n\ndid not find expected in actual.\n\nexpected: {expected_message}\nactual:{message}\n"
-                    );
-                    // errors should also contain  a help message
-                    assert!(message.contains("Must be a supported arrow type name such as 'Int32' or 'Timestamp(Nanosecond, None)'"));
-                }
-            }
-        }
-    }
+    val.parse().map_err(|e| match e {
+        arrow::error::ArrowError::ParseError(e) => DataFusionError::Plan(e),
```
Comment (comphead) [inline datafusion/functions/src/core/arrow_cast.rs:142]:
```suggestion
        arrow::error::ArrowError::ParseError(e) => plan_datafusion_error!(e),
```
?

File: datafusion/functions/src/core/arrow_cast.rs [hunk 4]
```diff
@@ -139,767 +137,9 @@ fn data_type_from_args(args: &[Expr]) -> Result<DataType> {
...
+        e => DataFusionError::ArrowError(e, None),
```
Comment (comphead) [inline datafusion/functions/src/core/arrow_cast.rs:143]:
```suggestion
        e => arrow_datafusion_error!(e),
```

**Review by comphead — APPROVED** [review]
lgtm thanks @alamb 
awesome clean up. errors macros can be added later

**Pull request merged**